source/sas.cpp

```diff
@@ -756,40 +758,89 @@ SAS::Timestamp SAS::Marker::get_timestamp() const
 }
 
 
-void SAS::log_to_stdout(log_level_t level,
-                        const char *module,
-                        int line_number,
-                        const char *fmt,
-                        ...)
+void SAS::sasclient_log_callback(log_level_t level,
+                                const char *module,
+                                int line_number,
+                                const char *fmt,
+                                ...)
 {
   va_list args;
-  const char* level_str;
-
   va_start(args, fmt);
+  _sasclient_log_callback(level, module, line_number, fmt, args);
+  va_end(args);
+}
 
-  switch (level) {
-    case LOG_LEVEL_ERROR:   level_str = "ERROR"; break;
-    case LOG_LEVEL_WARNING: level_str = "WARNING"; break;
-    case LOG_LEVEL_STATUS:  level_str = "STATUS"; break;
-    case LOG_LEVEL_INFO:    level_str = "INFO"; break;
-    case LOG_LEVEL_VERBOSE: level_str = "VERBOSE"; break;
-    case LOG_LEVEL_DEBUG:   level_str = "DEBUG"; break;
-    default:                level_str = "UNKNOWN"; break;
+
+// Write the internally raised log to a char array, then pass
+// this to the common logging callback to log.
+void SAS::_sasclient_log_callback(log_level_t level,
+                                  const char *module,
+                                  int line_number,
+                                  const char *fmt,
+                                  va_list args)
+{
+  const char* log_level[] = {"Error", "Warning", "Status", "Info", "Verbose", "Debug"};
+  char logline[MAX_LOGLINE];
+  int written = 0;
+
+  // Strip the directory from the file location.
+  const char* mod = strrchr(module, '/');
+  module = (mod != NULL) ? mod + 1 : module;
+
+  if (line_number)
+  {
+    written = snprintf(logline, MAX_LOGLINE - 2, "CHECK %s %s:%d: ", log_level[level], module, line_number);
+  }
+  else
+  {
+    written = snprintf(logline, MAX_LOGLINE - 2, "%s %s: ", log_level[level], module);
   }
 
-  printf("%s %s:%d: ", level_str, module, line_number);
-  vprintf(fmt, args);
-  printf("\n");
-  fflush(stdout);
+  // snprintf and vsnprintf return the bytes that would have been
+  // written if their second argument was large enough, so we need to
+  // reduce the size of written to compensate if it is too large.
+  written = std::min(written, MAX_LOGLINE - 2);
 
-  va_end(args);
-}
+  int bytes_available = MAX_LOGLINE - written - 2;
+  written += vsnprintf(logline + written, bytes_available, fmt, args);
 
+  if (written > (MAX_LOGLINE - 2))
```

Comment:
What is it always `MAX_LOGLINE - 2` Why the `-2`? Also why do you do `std::min(written, MAX_LOGLINE - 2);` above, but an if test here? Please can you be consistent?

Reply:
The -2 is there to leave space for the new line and null termination that gets added just before the array is handed over to the logger in log.cpp. I'll add a comment.

Reply:
Ah OK. NULL termination makes sense. But I don't think this function should be adding line breaks.

Reply:
It's not adding line breaks, but it's saving the space for when the logging framework needs to.

Reply:
But the logging framework is using a different buffer to sprintf into, right?

Reply:
Hmmmm if we give something too large to the logging framework it'll just truncate it as expected, which seems fine. But I have to declare an array of *some* size to sprintf into, so is it not sensible to make it the "correct" size?

Reply:
The logging framework can and will prepend some stuff onto the front of the log, so this function has no idea what the "right" size is.

I think my ultimate feedback is that this fucntion has too much responsibility. It should just be converting the parameters it receives into a format that allows it to call back into the user of the library. So all it needs to do is sprintf the parameters into a buffer and call `_log_callback`. Let the log library handle printing the level, newlines and whatever else it needs to. When viewed like this, it doesn't make sense for this function to second guess what the log framework is going to do.